### PR TITLE
[flash_ctrl] Adding test V1 Host Direct Read

### PR DESCRIPTION
**Pull request by nikola-miladinovic**
Signed-off-by: Nikola Miladinovic <nikola.miladinovic@ensilica.com>

### CODE REVIEW

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+        1 :/ cfg.seq_cfg.mp_region_program_en_pc
+      };
+
+      mp_regions[i].erase_en dist {
+        0 :/ (100 - cfg.seq_cfg.mp_region_erase_en_pc),
+        1 :/ cfg.seq_cfg.mp_region_erase_en_pc
+      };
+
+      mp_regions[i].start_page inside {[0:FlashNumPages - 1]};
+      mp_regions[i].num_pages inside {[1:FlashNumPages - mp_regions[i].start_page]};
```
Comment (tjaychen) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv:98]: does `FlashNumPages` here account for the multiple banks? Or is it max number of pages within 1 bank?

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+    `DV_CHECK_MEMBER_RANDOMIZE_FATAL(flash_op)
+    `DV_CHECK_MEMBER_RANDOMIZE_FATAL(flash_op_data)
+
+    `uvm_info(`gfn, $sformatf("Starting flash_ctrl op backdoor: %p", flash_op), UVM_LOW)
+    `uvm_info(`gfn, $sformatf("INIT BEFORE PROGRAM OP"), UVM_HIGH)
+    cfg.flash_mem_bkdr_write(.flash_op(flash_op), .scheme(FlashMemInitSet));
+
+    //setting half 1 due to FlashMemInitSet
+    exp_data = {{TL_DW/2{1'b0}},{TL_DW/2{1'b1}}};
```
Comment (tjaychen) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv:262]: is `FlashMemInitSet` the back door load?

File: hw/ip/flash_ctrl/dv/xfile [hunk 3]
```diff
@@ -0,0 +1 @@
+SCOPE tb.dut F
```
Comment (matutem) [inline hw/ip/flash_ctrl/dv/xfile:1]: Seems this file was added by mistake?

File: hw/ip/flash_ctrl/dv/env/flash_ctrl_env.core [hunk 4]
```diff
@@ -1,48 +1,49 @@
...
-      name: flash_ctrl
-      ip_hjson: ../../data/flash_ctrl.hjson
-
-targets:
-  default:
-    filesets:
-      - files_dv
-    generate:
-      - ral
+CAPI=2:
```
Comment (matutem) [inline hw/ip/flash_ctrl/dv/env/flash_ctrl_env.core:1]: Oops! it seems you are using windows, and have lines ending with the classical extra \r\n. Please use Unix line conventions.

See https://github.com/lowRISC/style-guides/blob/master/VerilogCodingStyle.md#posix-file-endings.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv [hunk 5]
```diff
@@ -0,0 +1,291 @@
+// Copyright lowRISC contributors.
```
Comment (matutem) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv:1]: This test should have a description in flash_ctrl/data/flash_ctrl_testplan.hjson. Please add it so we can understand what you are tying to test.

Comment (matutem) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv:1]: And this file also uses \r\n, so please change that.

**Review by matutem — CHANGES_REQUESTED** [review]
Please deal with the line endings and with a description in the test plan.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv (hunk 1 above)
Comment (nikola-miladinovic) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv:98]: > 
here it is:
parameter uint FlashNumPages            = flash_ctrl_pkg::NumBanks * flash_ctrl_pkg::PagesPerBank;

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv (hunk 2 above)
Comment (nikola-miladinovic) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv:262]: > 

Yes, backdoor is used.

File: hw/ip/flash_ctrl/dv/xfile (hunk 3 above)
Comment (nikola-miladinovic) [reply hw/ip/flash_ctrl/dv/xfile:1]: No, Xcelium does not work without following switch:
--build-opts "-xfile /projects/google/goo_opentitan/nikolam/opentitan/hw/ip/flash_ctrl/dv/xfile".
VCS works normal, but this switch is needed in order to override integrated XPROP in Xcelium. Otherwise Xcelium stays in infinite loop.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv (hunk 5 above)
Comment (nikola-miladinovic) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv:1]: Agreed. After merging #9025, i will update pull request. Also this test will be updated due to adding multiple RAL to Flash Ctrl.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv (hunk 2 above)
Comment (tjaychen) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv]: this recently merged PR #9416 may have an impact on your work.
just FYI.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv (hunk 1 above)
Comment (tjaychen) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv]: excellent, thank you.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+
+    `DV_CHECK_MEMBER_RANDOMIZE_FATAL(flash_op)
+    `DV_CHECK_MEMBER_RANDOMIZE_FATAL(flash_op_data)
+
+    `uvm_info(`gfn, $sformatf("Starting flash_ctrl op backdoor: %p", flash_op), UVM_LOW)
+    `uvm_info(`gfn, $sformatf("INIT BEFORE PROGRAM OP"), UVM_HIGH)
+    cfg.flash_mem_bkdr_write(.flash_op(flash_op), .scheme(FlashMemInitSet));
+
+    //setting half 1 due to FlashMemInitSet
+    exp_data = {{TL_DW/2{1'b0}},{TL_DW/2{1'b1}}};
```
Comment (tjaychen) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv:262]: quick question, will this test always test against this pattern? Or is the plan to enhance this separately later for random values?

Comment (nikola-miladinovic) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv:262]: This test will be completly rewritten using another RAL and with random values.for now,this test only read set values in order to prove that mem if works.

**Comment by nikola-miladinovic** [comment]
> Please deal with the line endings and with a description in the test plan.

Done.

**Review by tjaychen — APPROVED** [review]
looks good from my end.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv [hunk 7]
```diff
@@ -0,0 +1,291 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// basic smoke hw  if test vseq
```
Comment (cindychip) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv:5]: nit: super space.

File: hw/ip/flash_ctrl/dv/xfile (hunk 3 above)
Comment (cindychip) [inline hw/ip/flash_ctrl/dv/xfile:1]: Is this file needed?

**Review by cindychip — COMMENTED** [review]
LGTM with a few small suggestions.

File: hw/ip/flash_ctrl/dv/xfile (hunk 3 above)
Comment (nikola-miladinovic) [reply hw/ip/flash_ctrl/dv/xfile:1]: yes, https://github.com/lowRISC/opentitan/pull/9409#discussion_r759113296.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv (hunk 7 above)
Comment (nikola-miladinovic) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv:5]: all other seqs are written in that manner.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv (hunk 5 above)
Comment (nikola-miladinovic) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_smoke_hw_vseq.sv]: @matutem : added test description and name to testplan

**Comment by nikola-miladinovic** [comment]
Pull request #9604 has all changes from this pull request including smoke HW test with proper usage of RAL.
